src/librustc_mir/transform/validate.rs

```diff
@@ -77,4 +86,94 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
             }
         }
     }
+
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, _location: Location) {
+        match &terminator.kind {
+            TerminatorKind::Goto { target } => {
+                self.check_bb(terminator.source_info.span, *target);
+            }
+            TerminatorKind::SwitchInt { targets, .. } => {
```

Comment:
This could also check that `targets.len() == values.len() + 1` holds

Reply:
Nice idea, that even subsumes the current check. Done.